### ISSUE TITLE
Add http (socks) proxy support

### ISSUE THREAD

**Issue by zzz676** [ekutner/home-connect-hass#360]
Bosch servers have recently been unavailable for some countries for connection from this add-on. For this reason, there is a request to add the ability to use a proxy server to connect to Bosch servers.

**Comment by ekutner** [comment]
I don't remember that the appliances themselves have an option to configure a proxy, how do they connect to the HC servers in this case? How does the mobile app connect to the appliances when not connected to wifi?

**Comment by zzz676** [comment]
> I don't remember that the appliances themselves have an option to configure a proxy, how do they connect to the HC servers in this case? How does the mobile app connect to the appliances when not connected to wifi?

The devices themselves connect without problems, as does the proprietary Home Connect application. The problem, as I understand it, is precisely the connection through their API, through which the integration works.

**Comment by moskovskiy82** [comment]
> > I don't remember that the appliances themselves have an option to configure a proxy, how do they connect to the HC servers in this case? How does the mobile app connect to the appliances when not connected to wifi?
> 
> The devices themselves connect without problems, as does the proprietary Home Connect application. The problem, as I understand it, is precisely the connection through their API, through which the integration works.

Exactly. Furthermore a local only mode is supported by the HomeConenct. You can enable it in the app settings. See #353

**Comment by alexbilevskiy** [comment]
**TL;DR**
Set up VPN on your home router and redirect traffic for `api-rna.home-connect.com`, `singlekey-id.com` (both needed for auth flow) and `api.home-connect.com` (used by integration, ip address is the same as `api-rna...`) via vpn.

  

> How do you know the appliances connect to the HC cloud? 

From official home connect app (with wifi disabled)
![image](https://github.com/ekutner/home-connect-hass/assets/14160356/2a7b87e3-f7c6-4914-8ace-ce823acbc4a4)

> The whole scenario doesn't make sense to me, if HC are blocking some countries then the appliances should be blocked too, why would they only block API access but still allow the appliances to connect from the same countries?

Actually it's been like this for a while now. The funny thing is that even mobile app works fine (except authorization part), but not api.

> What error is generated when the integration is connecting to the API?

With vpn disabled, this error is shown
```
2024-04-08 19:09:53.884 ERROR (MainThread) [aiohttp_sse_client.client] fetch https://api.home-connect.com/api/homeappliances/events failed: 403
```
When requested with curl, there is 403 in plain html:
```
[alex@home ~]> curl https://api.home-connect.com/api/homeappliances/events
<html>
<head><title>403 Forbidden</title></head>
<body>
<center><h1>403 Forbidden</h1></center>
</body>
</html>
```
and this is with vpn 
```
[alex@home ~]> curl https://api.home-connect.com/api/homeappliances/events
{
  "error": {
    "key": "401",
    "description": "Authentication is possible but has failed or not yet been provided."
  }
}
```

**Comment by zzz676** [comment]
> **TL;DR** Set up VPN on your home router and redirect traffic for `api-rna.home-connect.com`, `singlekey-id.com` (both needed for auth flow) and `api.home-connect.com` (used by integration, ip address is the same as `api-rna...`) via vpn.
> 
> > Откуда вы знаете, что приборы подключаются к облаку HC?
> 
> Из официального приложения для подключения к дому (с отключенным Wi-Fi) ![Изображение](https://private-user-images.githubusercontent.com/14160356/320566986-2a7b87e3-f7c6-4914-8ace-ce823acbc4a4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2ODE5NTgsIm5iZiI6MTcxMjY4MTY1OCwicGF0aCI6Ii8xNDE2MDM1Ni8zMjA1NjY5ODYtMmE3Yjg3ZTMtZjdjNi00OTE0LThhY2UtY2U4MjNhY2JjNGE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDE2NTQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVmZGMyMGE3MzIzN2VjMDZkNTNhMmE5MmVjMDY3Y2Q3NGQ1MDM0N2RmMzUzNTE0YThkMDVmYjBlZmU4YzFlZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TK1apwxm0eU_BwOvBd-cPZ79WViVhTNLHgqmnpfZaZk)
> 
> > Весь сценарий не имеет смысла для меня, если HC блокирует некоторые страны, то приборы тоже должны быть заблокированы, почему они только блокируют доступ к API, но все же позволяют устройствам подключаться из тех же стран?
> 
> На самом деле, так было уже некоторое время. Забавно то, что даже мобильное приложение работает нормально (кроме части авторизации), но не api.
> 
> > Какая ошибка возникает при подключении интеграции к API?
> 
> При отключенном vpn эта ошибка отображается
> 
> ```
> 2024-04-08 19:09:53.884 ERROR (MainThread) [aiohttp_sse_client.client] fetch https://api.home-connect.com/api/homeappliances/events failed: 403
> ```
> 
> При запросе с помощью curl есть 403 в обычном html:
> 
> ```
> [alex@home ~]> curl https://api.home-connect.com/api/homeappliances/events
> <html>
> <head><title>403 Forbidden</title></head>
> <body>
> <center><h1>403 Forbidden</h1></center>
> </body>
> </html>
> ```
> 
> и это с VPN
> 
> ```
> [alex@home ~]> curl https://api.home-connect.com/api/homeappliances/events
> {
>   "error": {
>     "key": "401",
>     "description": "Authentication is possible but has failed or not yet been provided."
>   }
> }
> ```

Thank you for your reply. Couldn't answer, was unavailable. I confirm, everything is the same for me.

**Comment by zzz676** [comment]
> **TL;DR** Set up VPN on your home router and redirect traffic for `api-rna.home-connect.com`, `singlekey-id.com` (both needed for auth flow) and `api.home-connect.com` (used by integration, ip address is the same as `api-rna...`) via vpn.
> 
> > How do you know the appliances connect to the HC cloud?
> 
> From official home connect app (with wifi disabled) ![image](https://private-user-images.githubusercontent.com/14160356/320566986-2a7b87e3-f7c6-4914-8ace-ce823acbc4a4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2ODIyMjksIm5iZiI6MTcxMjY4MTkyOSwicGF0aCI6Ii8xNDE2MDM1Ni8zMjA1NjY5ODYtMmE3Yjg3ZTMtZjdjNi00OTE0LThhY2UtY2U4MjNhY2JjNGE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDE2NTg0OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ4NGI5MjAxYjQxZTkyNGZhZGIxNDZkM2I1YTBhNjIyMTViODRmNTNiODM1NjNjOTM1OTdjYzg4MjZjYjVlMjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.58C5VI4GYKOJfFG8enLQG75MUb9OhRbwY2WUvMMuDso)
> 
> > The whole scenario doesn't make sense to me, if HC are blocking some countries then the appliances should be blocked too, why would they only block API access but still allow the appliances to connect from the same countries?
> 
> Actually it's been like this for a while now. The funny thing is that even mobile app works fine (except authorization part), but not api.
> 
> > What error is generated when the integration is connecting to the API?
> 
> With vpn disabled, this error is shown
> 
> ```
> 2024-04-08 19:09:53.884 ERROR (MainThread) [aiohttp_sse_client.client] fetch https://api.home-connect.com/api/homeappliances/events failed: 403
> ```
> 
> When requested with curl, there is 403 in plain html:
> 
> ```
> [alex@home ~]> curl https://api.home-connect.com/api/homeappliances/events
> <html>
> <head><title>403 Forbidden</title></head>
> <body>
> <center><h1>403 Forbidden</h1></center>
> </body>
> </html>
> ```
> 
> and this is with vpn
> 
> ```
> [alex@home ~]> curl https://api.home-connect.com/api/homeappliances/events
> {
>   "error": {
>     "key": "401",
>     "description": "Authentication is possible but has failed or not yet been provided."
>   }
> }
> ```

Thank you for your reply. Couldn't answer, was unavailable. I confirm, everything is the same for me.
